python/ray/serve/_private/router.py

```diff
@@ -759,6 +800,39 @@ def _assign_replica(self, query: Query, replica: RunningReplicaInfo):
             self.in_flight_queries[replica].add(tracker_ref)
         return user_ref
 
+    def _check_valid_candidate(self, replica_id: str) -> bool:
+        """Checks if the replica is a valid candidate based on its code version."""
+        internal_replica_context = ray.serve.context.get_internal_replica_context()
+        if (
+            internal_replica_context is None
+            or internal_replica_context.replica_tag is None
+        ):
+            return True
+
+        current_app = internal_replica_context.app_name
+        # NOTE(zcin): This is a slightly hacky way of getting the application
+        # of the deployment to which this router sends requests. We include
+        # application info in RunningReplicaInfo, and fetch it from there.
+        # A better solution would be to separate deployment names from app
+        # names and have Serve handles take a deployment name and application
+        # name as input upon initialization.
+        target_app = {
+            replica_info.app_name for replica_info in self.in_flight_queries.keys()
+        }.pop()
```

Comment:
IMO the best solution for this is to completely separate deployment name and application name. However this requires replacing with (app_name, deployment_name) wherever deployment name is used as an identifier, and that is a pretty big change. Before we make that change, this is a hacky-ish solution for making sure this change doesn't restrict sending requests between applications. LMK if there are any other suggestions! (for either the temporary solution or the more-permanent solution)